lib/messageUtils.ts

```diff
@@ -131,12 +131,20 @@ export function getDashboard(
   );
   const dashboardId = getDashboardIdForTemplate(template);
 
+  let baseUrl = `https://mozilla.cloud.looker.com/dashboards/${dashboardId}`;
+
   if (_isAboutWelcomeTemplate(template)) {
-    return `https://mozilla.cloud.looker.com/dashboards/${dashboardId}?Submission+Timestamp+Date=${encodedSubmissionDate}&Message+ID=%25${encodedMsgId?.toUpperCase()}%25&Normalized+Channel=${encodedChannel}&Experiment=${encodedExperiment}&Branch=${encodedBranchSlug}`;
+    return new URL(
+      `?Submission+Timestamp+Date=${encodedSubmissionDate}&Message+ID=%25${encodedMsgId}%25%2C+%25${encodedMsgId.toLowerCase()}%25%2C+%25${encodedMsgId.toUpperCase()}%25&Normalized+Channel=${encodedChannel}&Experiment=${encodedExperiment}&Branch=${encodedBranchSlug}`,
```

Comment:
I think you did the right thing here based on what I communicated, and this does help validate stuff. That said, I didn't communicate what I had in mind at all clearly.

My thinking is to use not only `URL` but also `URLSearchParams` so that we don't do hand-encoding but focus entirely on getting the right parameters.

Here's a diff that shows a rough approximation of what I have in mind:

```diff
diff --git a/__tests__/lib/messageUtils.test.ts b/__tests__/lib/messageUtils.test.ts
index e5179b9..08a7abb 100644
--- a/__tests__/lib/messageUtils.test.ts
+++ b/__tests__/lib/messageUtils.test.ts
@@ -67,7 +67,7 @@ describe("toBinary", () => {
 describe("getDashboard", () => {
   it("returns a correct infobar dashboard link w/exp & branch", () => {
     const template = "infobar";
-    const msgId = "12`3"; // weird chars to test URI encoding
+    const msgId = "ab`c"; // weird chars to test URI encoding
     const channel = "release";
     const experiment = "experiment:test";
     const branchSlug = "treatment:a";
@@ -87,7 +87,7 @@ describe("getDashboard", () => {
 
   it("returns a correct featureCallout dashboard link", () => {
     const template = "feature_callout";
-    const msgId = "1:23"; // weird chars to test URI encoding
+    const msgId = "a:bc"; // weird chars to test URI encoding
     const dashboardId = getDashboardIdForTemplate(template);
 
     const expectedLink = `https://mozilla.cloud.looker.com/dashboards/${dashboardId}?Submission+Timestamp+Date=30%20day%20ago%20for%2030%20day&Message+ID=%25${encodeURIComponent(msgId)}%25%2C+%25${encodeURIComponent(msgId).toLowerCase()}%25%2C+%25${encodeURIComponent(msgId).toUpperCase()}%25&Normalized+Channel=&Experiment=&Branch=`;
@@ -99,7 +99,7 @@ describe("getDashboard", () => {
   // XXX should this be "about:welcome" to be consistent with featureIds?
   it("returns a correct aboutwelcome dashboard link w/exp & branch", () => {
     const template = "aboutwelcome";
-    const msgId = "1:23"; // weird chars to test URI encoding
+    const msgId = "a:bc"; // weird chars to test URI encoding
     const experiment = "experiment:test";
     const branchSlug = "treatment:a";
     const dashboardId = getDashboardIdForTemplate(template);
@@ -118,7 +118,7 @@ describe("getDashboard", () => {
 
   it("returns a correct dashboard link with defined start and end dates where the end date is in the future", () => {
     const template = "feature_callout";
-    const msgId = "1:23"; // weird chars to test URI encoding
+    const msgId = "a:bc"; // weird chars to test URI encoding
     const startDate = "2024-03-08";
     const endDate = "2025-06-28";
     const dashboardId = getDashboardIdForTemplate(template);
@@ -139,7 +139,7 @@ describe("getDashboard", () => {
 
   it("returns a correct dashboard link with defined start and end dates where the end date is in the past", () => {
     const template = "feature_callout";
-    const msgId = "1:23"; // weird chars to test URI encoding
+    const msgId = "a:bc"; // weird chars to test URI encoding
     const startDate = "2024-03-08";
     const endDate = "2024-05-28";
     const dashboardId = getDashboardIdForTemplate(template);
@@ -160,7 +160,7 @@ describe("getDashboard", () => {
 
   it("returns a correct dashboard link with a defined start date", () => {
     const template = "feature_callout";
-    const msgId = "1:23"; // weird chars to test URI encoding
+    const msgId = "a:bc"; // weird chars to test URI encoding
     const startDate = "2024-03-08";
     const dashboardId = getDashboardIdForTemplate(template);
 
diff --git a/lib/messageUtils.ts b/lib/messageUtils.ts
index 529c54f..e072211 100644
--- a/lib/messageUtils.ts
+++ b/lib/messageUtils.ts
@@ -119,25 +119,42 @@ export function getDashboard(
   endDate?: string | null,
   isCompleted?: boolean,
 ): string | undefined {
+
   const encodedMsgId = encodeURIComponent(msgId);
+  // XXX comment here about what we're trying to do here and why
+  // each of the three clauses here is required (assuming all three are)
+  const multiCaseMessageIds =
+    `%${msgId}%,` + `%${msgId.toLowerCase()}%,` + `%${msgId.toUpperCase()}%`;
+
   const encodedTemplate = encodeURIComponent(template);
   const encodedChannel = channel ? encodeURIComponent(channel) : "";
   const encodedExperiment = experiment ? encodeURIComponent(experiment) : "";
   const encodedBranchSlug = branchSlug ? encodeURIComponent(branchSlug) : "";
   // The isCompleted value can be useful for messages that used to be in remote
   // settings or old versions of Firefox.
-  const encodedSubmissionDate = encodeURIComponent(
-    getLookerSubmissionTimestampDateFilter(startDate, endDate, isCompleted),
+  const submissionDate = getLookerSubmissionTimestampDateFilter(
+    startDate,
+    endDate,
+    isCompleted,
   );
+  const encodedSubmissionDate = encodeURIComponent(submissionDate);
   const dashboardId = getDashboardIdForTemplate(template);
 
   let baseUrl = `https://mozilla.cloud.looker.com/dashboards/${dashboardId}`;
 
   if (_isAboutWelcomeTemplate(template)) {
-    return new URL(
-      `?Submission+Timestamp+Date=${encodedSubmissionDate}&Message+ID=%25${encodedMsgId}%25%2C+%25${encodedMsgId.toLowerCase()}%25%2C+%25${encodedMsgId.toUpperCase()}%25&Normalized+Channel=${encodedChannel}&Experiment=${encodedExperiment}&Branch=${encodedBranchSlug}`,
-      baseUrl,
-    ).toString();
+    const paramObj = {
+      "Submission Timestamp Date": submissionDate,
+      "Message ID": multiCaseMessageIds,
+      "Normalized Channel": channel ? channel : "",
+      Experiment: experiment ? encodeURIComponent(experiment) : "",
+      Branch: branchSlug ? branchSlug : "",
+    };
+    const params = new URLSearchParams(Object.entries(paramObj));
+
+    let n = new URL(baseUrl);
+    n.search = params.toString();
+    return n.toString();
   }
 
   if (template === "infobar") {
 ```
 
The changes to the tests here were just to make the output more readable so I could convince myself that I wasn't breaking things terribly.

Note that the tests are still failing, but I _think_ that the failures now are all valid variants and still work.  They also make the URLs easier to read which is a nice bonus. 

My current thinking is that we probably want the tests to also avoid dealing with the encoding stuff, but are likely to be reasonable if we switch them over to URL and its accessor functions/

Would love to hear thoughts from you and @AllegroFox on this as well.

Reply:
Ahh, I see. I agree using `URLSearchParams` makes a lot of sense and also makes it a lot more readable. And yes, after taking a look at the tests after using the `URLSearchParams`, I think that the failures is just due to the encoding being slightly different, but still valid. I'll try playing around with the tests so that the encoding should be consistent.